# A deep dive into TOB-WGS' pipeline

Chapter 1: what does `ReblockGVCF` actually do ? Let's examine sample TOB1520 (which is an outlier for some QC metrics)

**version 2.2 of the pipeline: `ReblockGVCF` now supposed NOT to remove low quality variants**

In [ ]:
import hail as hl;

# All datasets in TOB-WGS are using GRCh38
hl.init(default_reference='GRCh38');


Import TOB1520 GVCF and explore structure and content


In [10]:
gvcf = hl.import_vcf('gs://cpg-tob-wgs-test/gvcf/batch1/TOB1520.g.vcf.gz', min_partitions=12, force_bgz=True)
gvcf.describe()
gvcf.info.END.show()

----------------------------------------
Global fields:
    None
----------------------------------------
Column fields:
    's': str
----------------------------------------
Row fields:
    'locus': locus<GRCh38>
    'alleles': array<str>
    'rsid': str
    'qual': float64
    'filters': set<str>
    'info': struct {
        AS_InbreedingCoeff: array<float64>, 
        AS_QD: array<float64>, 
        AS_RAW_BaseQRankSum: str, 
        AS_RAW_MQ: str, 
        AS_RAW_MQRankSum: str, 
        AS_RAW_ReadPosRankSum: str, 
        AS_SB_TABLE: str, 
        BaseQRankSum: float64, 
        DP: int32, 
        DS: bool, 
        END: int32, 
        ExcessHet: float64, 
        InbreedingCoeff: float64, 
        MLEAC: array<int32>, 
        MLEAF: array<float64>, 
        MQRankSum: float64, 
        RAW_MQandDP: array<int32>, 
        ReadPosRankSum: float64
    }
----------------------------------------
Entry fields:
    'AD': array<int32>
    'DP': int32
    'GQ': int32
    'GT': call


2021-06-19 06:56:17 Hail: INFO: Coerced sorted dataset


+---------------+-------------------+--------+
| locus         | alleles           | <expr> |
+---------------+-------------------+--------+
| locus<GRCh38> | array<str>        |  int32 |
+---------------+-------------------+--------+
| chr1:10001    | ["T","<NON_REF>"] |  10016 |
| chr1:10017    | ["C","<NON_REF>"] |  10026 |
| chr1:10027    | ["A","<NON_REF>"] |  10027 |
| chr1:10028    | ["C","<NON_REF>"] |  10038 |
| chr1:10039    | ["A","<NON_REF>"] |  10041 |
| chr1:10042    | ["C","<NON_REF>"] |  10044 |
| chr1:10045    | ["A","<NON_REF>"] |  10045 |
| chr1:10046    | ["C","<NON_REF>"] |  10046 |
| chr1:10047    | ["C","<NON_REF>"] |  10047 |
| chr1:10048    | ["C","<NON_REF>"] |  10050 |
+---------------+-------------------+--------+
showing top 10 rows

In [11]:
hl.summarize_variants(gvcf)

2021-06-19 06:57:15 Hail: INFO: Coerced sorted dataset


==============================
Number of variants: 35550357
==============================
Alleles per variant
-------------------
  2 alleles: 30075685 variants
  3 alleles: 5202620 variants
  4 alleles: 235909 variants
  5 alleles: 26301 variants
  6 alleles: 6800 variants
  7 alleles: 1784 variants
  8 alleles: 1258 variants
==============================
Variants per contig
-------------------
   chr1: 2946558 variants
   chr2: 2798614 variants
   chr3: 2239708 variants
   chr4: 2158804 variants
   chr5: 1998760 variants
   chr6: 1963461 variants
   chr7: 2016563 variants
   chr8: 1651842 variants
   chr9: 1553560 variants
  chr10: 1675602 variants
  chr11: 1569278 variants
  chr12: 1658813 variants
  chr13: 1186838 variants
  chr14: 1083266 variants
  chr15: 1067085 variants
  chr16: 1189288 variants
  chr17: 1193802 variants
  chr18: 895233 variants
  chr19: 1047294 variants
  chr20: 874802 variants
  chr21: 515840 variants
  chr22: 624892 variants
   chrX: 1585249 variants
   chrY: 55205 variants
==============================
Allele type distribution
------------------------
   Symbolic: 35550357 alternate alleles
        SNP: 4464040 alternate alleles (Ti: 2747207, Tv: 1716833, ratio: 1.60)
  Insertion: 637685 alternate alleles
   Deletion: 591733 alternate alleles
       Star: 103550 alternate alleles
    Complex: 1 alternate alleles
==============================

Worth noting:

of 35,550,357 variants, 30,075,685 are homozygous reference blocks

Let's load the GVCF after `ReblockGVCF` had been performed

In [13]:
rb_gvcf = hl.import_vcf('gs://cpg-tob-wgs-test-tmp/joint-calling/v2.2/hail/batch/7c4ba7/1/output_gvcf.g.vcf.gz', force_bgz=True, min_partitions=12)
rb_gvcf.describe();
hl.summarize_variants(rb_gvcf);

----------------------------------------
Global fields:
    None
----------------------------------------
Column fields:
    's': str
----------------------------------------
Row fields:
    'locus': locus<GRCh38>
    'alleles': array<str>
    'rsid': str
    'qual': float64
    'filters': set<str>
    'info': struct {
        AC: array<int32>, 
        AF: array<float64>, 
        AN: int32, 
        AS_BaseQRankSum: array<float64>, 
        AS_FS: array<float64>, 
        AS_InbreedingCoeff: array<float64>, 
        AS_MQ: array<float64>, 
        AS_MQRankSum: array<float64>, 
        AS_QD: array<float64>, 
        AS_QUALapprox: str, 
        AS_RAW_BaseQRankSum: str, 
        AS_RAW_MQ: str, 
        AS_RAW_MQRankSum: str, 
        AS_RAW_ReadPosRankSum: str, 
        AS_ReadPosRankSum: array<float64>, 
        AS_SB_TABLE: str, 
        AS_SOR: array<float64>, 
        AS_VarDP: str, 
        BaseQRankSum: float64, 
        DP: int32, 
        DS: bool, 
        END: int32, 
   

2021-06-19 07:09:59 Hail: INFO: Coerced sorted dataset


==============================
Number of variants: 34953785
==============================
Alleles per variant
-------------------
  2 alleles: 29886034 variants
  3 alleles: 4824254 variants
  4 alleles: 211029 variants
  5 alleles: 23444 variants
  6 alleles: 6243 variants
  7 alleles: 1647 variants
  8 alleles: 1134 variants
==============================
Variants per contig
-------------------
   chr1: 2893335 variants
   chr2: 2753753 variants
   chr3: 2201870 variants
   chr4: 2125750 variants
   chr5: 1967506 variants
   chr6: 1933167 variants
   chr7: 1983526 variants
   chr8: 1627502 variants
   chr9: 1527980 variants
  chr10: 1646718 variants
  chr11: 1543450 variants
  chr12: 1630220 variants
  chr13: 1168333 variants
  chr14: 1065468 variants
  chr15: 1049219 variants
  chr16: 1167828 variants
  chr17: 1170796 variants
  chr18: 881781 variants
  chr19: 1025994 variants
  chr20: 857176 variants
  chr21: 506947 variants
  chr22: 611828 variants
   chrX: 1561403 variants
   chrY: 52235 variants
==============================
Allele type distribution
------------------------
   Symbolic: 34953785 alternate alleles
        SNP: 4135327 alternate alleles (Ti: 2701454, Tv: 1433873, ratio: 1.88)
  Insertion: 588737 alternate alleles
   Deletion: 543947 alternate alleles
       Star: 88643 alternate alleles
    Complex: 1 alternate alleles
==============================

These are the annotations added by ReblockGVCF

        AC, AF, AN: ???? why would we need that on a single sample ?
        AS_BaseQRankSum: array<float64>, 
        AS_FS: array<float64>,
        AS_MQ: array<float64>, 
        AS_MQRankSum: array<float64>,
        AS_QD: array<float64>, 
        AS_QUALapprox: str,
        AS_ReadPosRankSum: array<float64>,
        AS_SOR: array<float64>, 
        AS_VarDP: str,
        FS: float64, 
        MQ: float64,  
        MQ_DP: int32, 
        QD: float64, 
        QUALapprox: int32, 
        RAW_GT_COUNT: array<int32>, 
        SOR: float64, 
        VarDP: int32
and the annotations removed by ReblockGVCF

        MLEAC: array<int32>, 
        MLEAF: array<float64>, 

Additionally we lost more than 5e6 variants: not only homref blocks but also SNPs, etc.
Let's have a look at what we lost:

In [14]:
lost = gvcf.anti_join_rows(rb_gvcf.rows())
hl.summarize_variants(lost)

2021-06-19 07:15:38 Hail: INFO: Coerced sorted dataset
2021-06-19 07:16:36 Hail: INFO: Coerced sorted dataset


==============================
Number of variants: 687344
==============================
Alleles per variant
-------------------
  2 alleles: 280423 variants
  3 alleles: 378366 variants
  4 alleles: 24880 variants
  5 alleles: 2857 variants
  6 alleles: 557 variants
  7 alleles: 137 variants
  8 alleles: 124 variants
==============================
Variants per contig
-------------------
   chr1: 61859 variants
   chr2: 51571 variants
   chr3: 43550 variants
   chr4: 37815 variants
   chr5: 35760 variants
   chr6: 34553 variants
   chr7: 38132 variants
   chr8: 27709 variants
   chr9: 29387 variants
  chr10: 33361 variants
  chr11: 29478 variants
  chr12: 32663 variants
  chr13: 21336 variants
  chr14: 20400 variants
  chr15: 20566 variants
  chr16: 24867 variants
  chr17: 26661 variants
  chr18: 15529 variants
  chr19: 24382 variants
  chr20: 20897 variants
  chr21: 10383 variants
  chr22: 15421 variants
   chrX: 27324 variants
   chrY: 3740 variants
==============================
Allele type distribution
------------------------
   Symbolic: 687344 alternate alleles
        SNP: 328713 alternate alleles (Ti: 45753, Tv: 282960, ratio: 0.16)
  Insertion: 48948 alternate alleles
   Deletion: 47786 alternate alleles
       Star: 14907 alternate alleles
==============================

We lost mostly homref blocks, as expected, but also other kind of variants, such as 425,204 SNPs

Let's inspect some of these lost variants (more than 2 alleles means it is not a homref block)

In [15]:
lost.filter_rows(hl.len(lost.alleles) == 3).show(50,1,truncate=30)

2021-06-19 07:19:09 Hail: INFO: Coerced sorted dataset
2021-06-19 07:20:05 Hail: INFO: Coerced sorted dataset
2021-06-19 07:21:03 Hail: INFO: Coerced sorted dataset


+---------------+--------------------------------+--------------+--------------+
| locus         | alleles                        | 'TOB1520'.AD | 'TOB1520'.DP |
+---------------+--------------------------------+--------------+--------------+
| locus<GRCh38> | array<str>                     | array<int32> |        int32 |
+---------------+--------------------------------+--------------+--------------+
| chr1:10109    | ["AACCCT","A","<NON_REF>"]     | [11,1,0]     |           12 |
| chr1:10231    | ["CCCCTAACCCTAACCCTAAACCCTA... | [8,0,0]      |            8 |
| chr1:10441    | ["CCCTA","C","<NON_REF>"]      | [11,0,0]     |           11 |
| chr1:16571    | ["G","A","<NON_REF>"]          | [33,2,0]     |           35 |
| chr1:16688    | ["G","A","<NON_REF>"]          | [20,2,0]     |           22 |
| chr1:19226    | ["T","A","<NON_REF>"]          | [27,2,0]     |           29 |
| chr1:49291    | ["C","T","<NON_REF>"]          | [11,1,0]     |           12 |
| chr1:58188    | ["G","T","<NON_REF>"]          | [8,1,0]      |            9 |
| chr1:83965    | ["AAAG","A","<NON_REF>"]       | [4,0,0]      |            4 |
| chr1:99066    | ["TTTC","T","<NON_REF>"]       | [10,2,0]     |           12 |
| chr1:99088    | ["C","G","<NON_REF>"]          | [8,1,0]      |            9 |
| chr1:109262   | ["C","A","<NON_REF>"]          | [24,1,0]     |           25 |
| chr1:183662   | ["G","C","<NON_REF>"]          | [126,13,0]   |          139 |
| chr1:186765   | ["T","C","<NON_REF>"]          | [33,3,0]     |           36 |
| chr1:187204   | ["G","A","<NON_REF>"]          | [31,2,0]     |           33 |
| chr1:189955   | ["A","G","<NON_REF>"]          | [31,2,0]     |           33 |
| chr1:190667   | ["G","A","<NON_REF>"]          | [66,6,0]     |           72 |
| chr1:191859   | ["C","T","<NON_REF>"]          | [64,6,0]     |           70 |
| chr1:197628   | ["G","C","<NON_REF>"]          | [17,2,0]     |           19 |
| chr1:198861   | ["G","C","<NON_REF>"]          | [36,3,0]     |           39 |
| chr1:198872   | ["C","T","<NON_REF>"]          | [34,3,0]     |           37 |
| chr1:261438   | ["A","G","<NON_REF>"]          | [33,3,0]     |           36 |
| chr1:269963   | ["T","A","<NON_REF>"]          | [11,1,0]     |           12 |
| chr1:272389   | ["C","T","<NON_REF>"]          | [41,5,0]     |           46 |
| chr1:275513   | ["A","G","<NON_REF>"]          | [34,2,0]     |           36 |
| chr1:277901   | ["C","A","<NON_REF>"]          | [27,1,0]     |           28 |
| chr1:284892   | ["G","A","<NON_REF>"]          | [37,4,0]     |           41 |
| chr1:285416   | ["T","G","<NON_REF>"]          | [24,2,0]     |           26 |
| chr1:286145   | ["C","CAT","<NON_REF>"]        | [28,2,0]     |           30 |
| chr1:455086   | ["C","A","<NON_REF>"]          | [15,2,0]     |           17 |
| chr1:590492   | ["TA","T","<NON_REF>"]         | [8,0,0]      |            8 |
| chr1:590500   | ["T","C","<NON_REF>"]          | [8,0,0]      |            8 |
| chr1:601557   | ["T","G","<NON_REF>"]          | [17,1,0]     |           18 |
| chr1:601846   | ["T","G","<NON_REF>"]          | [12,2,0]     |           14 |
| chr1:602196   | ["TCCTCTCCTTTCTTCTCTCCATCCC... | [8,0,0]      |            8 |
| chr1:667265   | ["C","T","<NON_REF>"]          | [19,2,0]     |           21 |
| chr1:690012   | ["T","C","<NON_REF>"]          | [20,2,0]     |           22 |
| chr1:712244   | ["G","T","<NON_REF>"]          | [9,1,0]      |           10 |
| chr1:737292   | ["C","T","<NON_REF>"]          | [36,2,0]     |           38 |
| chr1:737436   | ["A","G","<NON_REF>"]          | [41,3,0]     |           44 |
| chr1:757165   | ["A","T","<NON_REF>"]          | [8,3,0]      |           11 |
| chr1:769039   | ["A","ACCCCCCCCCCCCCCCCCCCC... | [9,0,0]      |            9 |
| chr1:780612   | ["TA","T","<NON_REF>"]         | [16,0,0]     |           16 |
| chr1:789646   | ["T","G","<NON_REF>"]          | [21,2,0]     |           23 |
| chr1:789804   | ["G","T","<NO

In [16]:
lost = lost.annotate_rows(nonhomref=hl.agg.count_where(lost.GT.is_non_ref())>0)
lost.aggregate_rows(hl.agg.count_where(lost.nonhomref))

2021-06-19 07:22:01 Hail: INFO: Coerced sorted dataset
2021-06-19 07:22:59 Hail: INFO: Coerced sorted dataset


0

**GOOD, no non HomRef variants were removed**